solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -642,8 +647,10 @@ private void constructState(Set<String> changedCollections) {
 
   /** Refresh collections. */
   private void refreshCollections() {
-    for (String coll : collectionWatches.watchedCollections()) {
-      new StateWatcher(coll).refreshAndWatch();
```

Comment:
Good catch on this @magibney !

I traced the code logic and design a little bit. The principle design for removal of watches (ie NOT re-install the watch to ZK after event trigger) appears to tie to the absence of collection name in `collectionWatches.watchedCollections()` in `StateWatcher#process` (which you pointed out later)

IMHO, this new StateWatcher() call is a bit of an oddball, it skips the checks on `collectionWatches` (unlike 2 other places that instantiate the `StateWatcher`, which check `collectionWatches` and only create new instance if there's no entry for the target collection), hence this StateWatcher fails to remove itself at https://github.com/cowpaths/fullstory-solr/pull/151/files#diff-9b1918d641f78a8c245306b47ffd7a4554a12771ed4cb42293a411537fe7b54dL1334.

I cannot find documentation stating that there could ever be only one `StateWatcher` instance per collection per zkReader - or perhaps I missed it, but such could have been implied by the current design of `collectionWatches`.

I'm unsure of the expected behaviors of this `refreshCollections` method (javadoc seems lacking 😅 ), it might simply want to re-fetch the collection states instead of installing a new zk watch? It could be using `new StateWatcher(coll).refreshAndWatch();` out of convenience, but it installs an extra zk watch for a collection already exists in `collectionWatches`.

Or if this also wants to install a new watch, then i think @magibney 's solution makes perfect sense - instead of having only `collectionWatches` field to keep track of what collections have zk watch installed, we also need to know exactly which `StateWatcher` instance is being used.